README.md
# NAME

SQL::Translator::Producer::GraphQL - GraphQL schema producer for SQL::Translator

# PROJECT STATUS

| OS      |  Build status |
|:-------:|--------------:|
| Linux   | [![Build Status](https://travis-ci.org/graphql-perl/SQL-Translator-Producer-GraphQL.svg?branch=master)](https://travis-ci.org/graphql-perl/SQL-Translator-Producer-GraphQL) |

[![CPAN version](https://badge.fury.io/pl/SQL-Translator-Producer-GraphQL.svg)](https://metacpan.org/pod/SQL::Translator::Producer::GraphQL)

# SYNOPSIS

    use SQL::Translator;
    use SQL::Translator::Producer::GraphQL;
    my $t = SQL::Translator->new( parser => '...' );
    $t->producer('GraphQL');
    $t->translate;

# DESCRIPTION

This module will produce a [GraphQL::Schema](https://metacpan.org/pod/GraphQL::Schema) from the given
[SQL::Translator::Schema](https://metacpan.org/pod/SQL::Translator::Schema). It does this by first
turning it into a [DBIx::Class::Schema](https://metacpan.org/pod/DBIx::Class::Schema) using
[SQL::Translator::Producer::DBIx::Class::File](https://metacpan.org/pod/SQL::Translator::Producer::DBIx::Class::File), and introspecting it.

Its `Query` type represents a guess at what fields are suitable, based
on providing a lookup for each type (a [DBIx::Class::ResultSource](https://metacpan.org/pod/DBIx::Class::ResultSource))
by each of its columns.

The `Mutation` type is similar: one `create/update/delete(type)` per
"real" type.

# ARGUMENTS

Currently none.

# EXPORTS

## schema\_dbic2graphql

Takes as input a [DBIx::Class::Schema](https://metacpan.org/pod/DBIx::Class::Schema) object, returns a
[GraphQL::Schema](https://metacpan.org/pod/GraphQL::Schema) object. E.g.:

    perl -MSQL::Translator::Producer::GraphQL=schema_dbic2graphql \
      -MModule::Runtime=require_module \
      -e '
        my $dbic_class = shift;
        require_module $dbic_class;
        print schema_dbic2graphql($dbic_class->connect)->to_doc;
      ' \
      -It/lib-dbicschema Schema | less

# DEBUGGING

To debug, set environment variable `GRAPHQL_DEBUG` to a true value.

# AUTHOR

Ed J, `<etj at cpan.org>`

Based heavily on [SQL::Translator::Producer::DBIxSchemaDSL](https://metacpan.org/pod/SQL::Translator::Producer::DBIxSchemaDSL).

# LICENSE

Copyright (C) Ed J

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension SQL-Translator-Producer-GraphQL

0.02	Wed Nov  1 04:13:43 GMT 2017
 - move translation to GraphQL::Plugin::Convert::DBIC, use that

0.01	Fri Oct 27 06:49:51 BST 2017
 - original version


lib/SQL/Translator/Producer/GraphQL.pm
package SQL::Translator::Producer::GraphQL;
use 5.008001;
use strict;
use warnings;
use SQL::Translator::Producer::DBIx::Class::File;
use GraphQL::Plugin::Convert::DBIC;

our $VERSION = "0.02";

my $dbic_schema_class_track = 'CLASS00000';
sub produce {
  my $translator = shift;
  my $schema = $translator->schema;
  my $dbic_schema_class = ++$dbic_schema_class_track;
  my $dbic_translator = bless { %$translator }, ref $translator;
  $dbic_translator->producer_args({ prefix => $dbic_schema_class });
  eval SQL::Translator::Producer::DBIx::Class::File::produce($dbic_translator);
  die "Failed to make DBIx::Class::Schema: $@" if $@;
  my $converted = GraphQL::Plugin::Convert::DBIC->to_graphql(
    sub { $dbic_schema_class->connect }
  );
  $converted->{schema}->to_doc;
}

=encoding utf-8

=head1 NAME

SQL::Translator::Producer::GraphQL - GraphQL schema producer for SQL::Translator

=begin markdown

# PROJECT STATUS

| OS      |  Build status |
|:-------:|--------------:|
| Linux   | [![Build Status](https://travis-ci.org/graphql-perl/SQL-Translator-Producer-GraphQL.svg?branch=master)](https://travis-ci.org/graphql-perl/SQL-Translator-Producer-GraphQL) |

[![CPAN version](https://badge.fury.io/pl/SQL-Translator-Producer-GraphQL.svg)](https://metacpan.org/pod/SQL::Translator::Producer::GraphQL)

=end markdown

=head1 SYNOPSIS

  use SQL::Translator;
  use SQL::Translator::Producer::GraphQL;
  my $t = SQL::Translator->new( parser => '...' );
  $t->producer('GraphQL');
  $t->translate;

=head1 DESCRIPTION

This module will produce a L<GraphQL::Schema> from the given
L<SQL::Translator::Schema>. It does this by first
turning it into a L<DBIx::Class::Schema> using
L<SQL::Translator::Producer::DBIx::Class::File>, then passing it to
L<GraphQL::Plugin::Convert::DBIC/to_graphql>.

=head1 ARGUMENTS

Currently none.

=head1 DEBUGGING

To debug, set environment variable C<GRAPHQL_DEBUG> to a true value.

=head1 AUTHOR

Ed J, C<< <etj at cpan.org> >>

Based heavily on L<SQL::Translator::Producer::DBIxSchemaDSL>.

=head1 LICENSE

Copyright (C) Ed J

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;


Makefile.PL
use 5.008001;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
  NAME             => 'SQL::Translator::Producer::GraphQL',
  AUTHOR           => q{Ed J <etj@cpan.org>},
  VERSION_FROM     => 'lib/SQL/Translator/Producer/GraphQL.pm',
  ABSTRACT_FROM    => 'lib/SQL/Translator/Producer/GraphQL.pm',
  LICENSE          => 'artistic_2',
  MIN_PERL_VERSION => '5.014',
  CONFIGURE_REQUIRES => {
    'ExtUtils::MakeMaker' => '0',
  },
  TEST_REQUIRES => {
    'Test::More' => '0.98',
    'SQL::Translator' => '0',
    'SQL::Translator::Producer::DBIx::Class::File' => '0',
    'GraphQL' => '0.20', # convert plugin
    'GraphQL::Plugin::Convert::DBIC' => '0.01',
  },
  PREREQ_PM => {
    'SQL::Translator' => '0',
    'SQL::Translator::Producer::DBIx::Class::File' => '0',
    'GraphQL' => '0.20', # convert plugin
    'GraphQL::Plugin::Convert::DBIC' => '0.01',
  },
  clean => { FILES => 'SQL-Translator-Producer-GraphQL-*' },
  META_MERGE => {
    "meta-spec" => { version => 2 },
    dynamic_config => 0,
    resources => {
      x_IRC => 'irc://irc.perl.org/#graphql-perl',
      repository => {
        type => 'git',
        url => 'git@github.com:graphql-perl/SQL-Translator-Producer-GraphQL.git',
        web => 'https://github.com/graphql-perl/SQL-Translator-Producer-GraphQL',
      },
      bugtracker  => {
        web => 'https://github.com/graphql-perl/SQL-Translator-Producer-GraphQL/issues',
      },
      license => [ 'http://dev.perl.org/licenses/' ],
    },
    prereqs => {
      develop => {
        requires => {
          'Test::CheckManifest' => '0.9',
          'Test::Pod' => '1.22',
          'Pod::Markdown' => 0,
        },
      },
    },
  },
);

sub MY::postamble {
  <<EOF;
pure_all :: README.md

README.md : lib/SQL/Translator/Producer/GraphQL.pm
\tpod2markdown \$< >\$\@
EOF
}
